### PR TITLE
Different fixes before release of online workspaces

### PR DESCRIPTION
**Pull request by RaynalHugo**
# Feature

## Work performed

- Update env files to provide `VERCEL_URL` rather than `NEXTAUH_URL`
- Implemented optimistic response on some effects mutations
- Updated some mutation return payload to benefit from Apollo's automated cache updates
- Replaced some refetchQueries with cache updates
- deduped some queries to reduce the number of active queries Apollo has to maintain

## Results

- Sending emails should work again :)
- A lot of interactions are now way faster on slow connections

## Problems encountered

There is still a lot of room for enhancements. 
- There are a lot of duplicated queries. We could investigate to fetch data in smarter ways.
- Some mutations still need to be performed sequentially (create a new label class before updating a label with the newly created class). The optimistic response isn't totaly useful here. It only reduces half of the waiting time as it needs to await the first mutation to prevent a race condition
- I left some refetchQueries about some counts as it was too difficult for now to properly implement the cache updates.

## Newly raised issues

Not newly but #522 would be a good follow-up issue for further investigations.

### CODE REVIEW

**Comment by nicOwlas** [comment]
@RaynalHugo I have this error when testing on above preview url (service worker does not load)

image

**Comment by RaynalHugo** [comment]
> @RaynalHugo I have this error when testing on above preview url (service worker does not load)

Strange, it worked perfectly one the previous deployment :/ But now I can reproduce the error. 

I've just triggered another deployment. Let's see if it persists.

**Comment by RaynalHugo** [comment]
@nicOwlas it seems it was a deployment issue, it now works on my side.

File: typescript/web/src/connectors/undo-store/effects/create-label.ts [hunk 1]
```diff
@@ -160,76 +28,52 @@ export const createCreateLabelEffect = (
...
       },
     });
 
-    if (typeof data?.createLabel?.id !== "string") {
-      throw new Error("Couldn't get the id of the newly created label");
-    }
+    // TODO: Ideally we could select the label before awaiting for the mutation to complete
+    // because the optimistic response is run synchronously.
+    // However it makes some query fails. It would be nice to try to reverse those 2 lines
+    // and make the other "cache-first"
```
Comment (etiennedupont) [inline typescript/web/src/connectors/undo-store/effects/create-label.ts:59]: Isn't there a way to select the label "inside" the optimistic response to ensure the queries don't fail?

**Review by etiennedupont — APPROVED** [review]
Some comments / questions but looks good to me!

File: typescript/web/src/connectors/apollo-client/client.ts [hunk 2]
```diff
@@ -2,6 +2,7 @@ import { ApolloClient, InMemoryCache, HttpLink, concat } from "@apollo/client";
 import { awaitServiceWorkerLink } from "./await-service-worker-link";
 
 export const serviceWorkerClient = new ApolloClient({
+  connectToDevTools: false,
```
[the remark this answers is not in the document]
Comment (RaynalHugo) [reply typescript/web/src/connectors/apollo-client/client.ts:5]: We have 2 different apollo clients, one for the Service Worker backend, and one for the distant DB backend. You can only connect one of them to the dev tools. The Service Worker one was connected by default and that's not the most interesting one.

File: typescript/web/src/connectors/undo-store/effects/create-label-class-and-update-label.ts [hunk 3]
```diff
@@ -67,43 +32,74 @@ export const createCreateLabelClassAndUpdateLabelEffect = (
...
-    const labelClassIdPrevious = labelClass?.id ?? null;
+    /** There is room for enhancement here.
+     * We await the mutation that create the label class before triggering the mutation which updates the label with the new label class.
+     * This makes the optimistic less useful.
+     * We still need to wait for the first mutation to answer before doing the update and optimistic response of the second one.
+     *
+     * Technically, it is not needed to wait for the first mutation as we already now the id of the label class we will create, however,
+     * we have a check in the backend to ensure that the new label class exist before updating a label. This would add a race condition.
+     *
+     * It would be better if we could create the label class and update the label in the same mutation.
```
[the remark this answers is not in the document]
Comment (RaynalHugo) [reply typescript/web/src/connectors/undo-store/effects/create-label-class-and-update-label.ts:51]: Do you mean, on the backend side? On the frontend side, I don't think that the "transaction" concept exists.

But on the backend side, we can await the first mutation before doing the second one. 
Ideally, with Prisma, we would be able to do something like `await prisma.label.update({labelClass: {create: {class:"new"}}})`

File: typescript/web/src/connectors/undo-store/effects/create-label.ts (hunk 1 above)
Comment (RaynalHugo) [reply typescript/web/src/connectors/undo-store/effects/create-label.ts:59]: That's what I wanted to do at first. 

However, it triggers re-renders in some parts of the UI.
And those parts triggers new queries looking for the details of the selected label. However, the select label doesn't yet exist while the mutation is pending and those queries fail... (I thought the race condition would be negligible but it was happening almost every time).

We should make those queries cache-first and ensure that everything they query is already available in the cache. Then, we will be able to change that.

File: typescript/web/src/connectors/undo-store/effects/create-label-class-and-update-label.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (etiennedupont) [reply typescript/web/src/connectors/undo-store/effects/create-label-class-and-update-label.ts:51]: I meant like executing two mutations at once like below. Does that make sense?
```graphql
mutation createLabelClassAndUpdateLabel{
mutation createLabelClass{
...
}
mutation updateLabel{
...
}
}
```

[the remark this answers is not in the document]
Comment (RaynalHugo) [reply typescript/web/src/connectors/undo-store/effects/create-label-class-and-update-label.ts:51]: Ah, I see. Unfortunately, this syntax isn't supported by apollo.

[the remark this answers is not in the document]
Comment (crubier) [reply typescript/web/src/connectors/undo-store/effects/create-label-class-and-update-label.ts:51]: This should be supported by Apollo no ?

```
mutation createLabelClassAndUpdateLabel{
  createLabelClass{
    ...
  }
  updateLabel{
    ...
  }
}
```

**Pull request merged**